### PR TITLE
AJAX-paginate Message Log report

### PR DESCRIPTION
**Pull request by kaapstorm**
[FB 180161](http://manage.dimagi.com/default.asp?180161)

(Don't worry about that PEP8 commit. That was just to feed my OCD monster.  The important stuff is in commit 73873fc.)

@gcapalbo , @snopoke any hints on the best way to test this change?

Holding off for test.

### CODE REVIEW

**Comment by kaapstorm** [comment]
Thanks @snopoke. Amended commit with changes.

**Comment by snopoke** [comment]
Looks good. I think as long as you were able to test it locally it should be OK, not sure if you have any sms messages in your local DB though.

**Comment by gcapalbo** [comment]
The changes look fine, thanks for the pep8 update (some of that code was pretty old).

But when I run this locally the report rows don't get generated and I instead get this error "There was an error with your query, it has been logged, please try another query."  The exception seems to be raised from the call to `json_format_datetime` that's used to actually render the json response (not the new calls to `json_format_datetime` that were added)

Also, does this still handle sorting and filtering?  And is the download affected by the pagination as well, or does the download option still let you download everything in the date range (we'd want to still be able to download everything)?

(Let me know if you need help generating local sms content for testing)

**Comment by kaapstorm** [comment]
OK, switched to using SQL data. Pagination works. We can sort by all columns except the message types column, because the value comes from different fields and then concatenated. Message type filters seem to work, but might need some QA because in converting to a SQL query I had to rethink how we filter them. I don't like how the filter function turned out, but I'm not sure how to simplify it; I'm wide open to suggestions.

File: corehq/apps/reports/standard/sms.py [hunk 1]
```diff
@@ -322,49 +319,123 @@ def uses_locations(self):
...
+            filtered_types = {t.lower() for t in MessageTypeFilter.get_value(self.request, self.domain)}
+            if not filtered_types:
+                return data_
+
+            relevant_workflows = (
+                WORKFLOW_REMINDER,
+                WORKFLOW_KEYWORD,
+                WORKFLOW_BROADCAST,
+                WORKFLOW_CALLBACK,
+                WORKFLOW_DEFAULT,
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: you'll want to merge with master and add `WORKFLOW_PERFORMANCE` here

File: corehq/apps/reports/standard/sms.py [hunk 2]
```diff
@@ -322,49 +319,123 @@ def uses_locations(self):
...
+                return phone_number[0:7] if phone_number[0] == "+" else phone_number[0:6]
+            return phone_number
+
+        get_direction = lambda d: {INCOMING: _('Incoming'), OUTGOING: _('Outgoing')}.get(d, '---')
+
+        def get_timestamp(date_):
+            timestamp = ServerTime(date_).user_time(self.timezone).done()
+            return timestamp.strftime(SERVER_DATETIME_FORMAT)
+
+        def get_contact_link(couch_recipient, couch_recipient_doc_type):
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: why not use `_fmt_contact_link` here?

File: corehq/apps/reports/standard/sms.py [hunk 3]
```diff
@@ -322,49 +319,123 @@ def uses_locations(self):
...
+            domain=self.domain,
+            date__range=(startdate, enddate),
+        ).exclude(
+            # Exclude outgoing messages that have not yet been processed
+            direction=OUTGOING,
+            processed=False
+        )
+        data = filter_by_types(data)
+        data = order_by_col(data)
+        data = data[self.pagination.start:self.pagination.start + self.pagination.count]
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: is there a way to do the pagination without querying for the whole result set first?

File: corehq/apps/reports/standard/sms.py [hunk 4]
```diff
@@ -322,49 +319,123 @@ def uses_locations(self):
 
     @property
     def rows(self):
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: is this going to mean that when you download this report, it will give you only the paginated records, or can you still download the entire date range?

File: corehq/apps/reports/standard/sms.py [hunk 5]
```diff
@@ -322,49 +319,123 @@ def uses_locations(self):
...
+        def get_phone_number(phone_number):
+            if abbreviate_phone_number and phone_number is not None:
+                return phone_number[0:7] if phone_number[0] == "+" else phone_number[0:6]
+            return phone_number
+
+        get_direction = lambda d: {INCOMING: _('Incoming'), OUTGOING: _('Outgoing')}.get(d, '---')
+
+        def get_timestamp(date_):
+            timestamp = ServerTime(date_).user_time(self.timezone).done()
+            return timestamp.strftime(SERVER_DATETIME_FORMAT)
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: why not use `_fmt_timestamp` to do the formatting?

File: corehq/apps/reports/standard/sms.py [hunk 6]
```diff
@@ -322,49 +319,123 @@ def uses_locations(self):
...
+            if self.request_params.get('sSortDir_0') == 'desc':
+                data_ = data_.reverse()
+            return data_
+
+        def get_phone_number(phone_number):
+            if abbreviate_phone_number and phone_number is not None:
+                return phone_number[0:7] if phone_number[0] == "+" else phone_number[0:6]
+            return phone_number
+
+        get_direction = lambda d: {INCOMING: _('Incoming'), OUTGOING: _('Outgoing')}.get(d, '---')
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: why not use `_fmt_direction`?

File: corehq/apps/reports/standard/sms.py (hunk 2 above)
Comment (kaapstorm) [inline corehq/apps/reports/standard/sms.py]: Good call. _fmt_contact_link returns a dictionary but we can just pull out the value we want.

File: corehq/apps/reports/standard/sms.py (hunk 3 above)
Comment (kaapstorm) [inline corehq/apps/reports/standard/sms.py]: In Django [QuerySets are lazy][1], and are only executed when they are iterated, so slicing the unevaluated QuerySet, as we are doing here, won't evaluate it yet. It will just add "OFFSET" and "LIMIT" clauses to the query string.

  [1]: https://docs.djangoproject.com/en/1.8/ref/models/querysets/

File: corehq/apps/reports/standard/sms.py (hunk 4 above)
Comment (kaapstorm) [inline corehq/apps/reports/standard/sms.py]: Thanks. Now checking whether to paginate first.

**Comment by kaapstorm** [comment]
Bump.

File: corehq/apps/reports/standard/sms.py [hunk 7]
```diff
@@ -324,48 +334,141 @@ def uses_locations(self):
...
+        startdate = json_format_datetime(self.datespan.startdate_utc)
+        enddate = json_format_datetime(self.datespan.enddate_utc)
+        result = SMSLog.view(
+            "sms/by_domain",
+            reduce=True,
+            startkey=[self.domain, 'SMSLog', startdate],
+            endkey=[self.domain, 'SMSLog', enddate],
+            include_docs=False
+        ).first()
+        return result['value'] if result else 0
```
Comment (snopoke) [inline corehq/apps/reports/standard/sms.py]: I think we should use SQL here as well so that its consistent.

File: corehq/apps/sms/models.py [hunk 8]
```diff
@@ -169,14 +169,22 @@ def count_outgoing_by_domain(cls, domain, start_date = None, end_date = {}):
...
         return cls.view("sms/by_domain",
-                    reduce=False,
-                    startkey=[domain, cls.__name__] + [start_date],
-                    endkey=[domain, cls.__name__] + [end_date],
-                    include_docs=True)
+                        reduce=False,
+                        startkey=[domain, cls.__name__] + [start_date],
+                        endkey=[domain, cls.__name__] + [end_date],
+                        include_docs=True,
+                        **kwargs)
```
Comment (snopoke) [inline corehq/apps/sms/models.py]: This doesn't seem necessary anymore now that we've switched to SQL

File: corehq/apps/reports/standard/sms.py [hunk 9]
```diff
@@ -170,7 +167,18 @@ def _fmt_contact_link(self, recipient_id, doc_info, extra_text=None):
...
+            # table_cell['raw'] available for export. export_table will read contacttype and recipientid
+            # attributes to populate those columns in the export spreadsheet.
+            ret = self.table_cell(
+                username,
+                '<a target="_blank" href="{url}" contacttype="{type}" recipientid="{id}">{username}</a>'.format(
+                    username=username,
+                    url=url,
+                    type=contact_type,
+                    id=recipient_id or ''
+                ))
```
Comment (snopoke) [inline corehq/apps/reports/standard/sms.py]: I'm not sure about this approach. If you look at `GenericTabularReport.export_table` you'll see that it calls `GenericTabularReport.export_rows` which you can override and add in additional formatting etc. In `export_rows` you do have access to `cell['raw']`

Comment (kaapstorm) [inline corehq/apps/reports/standard/sms.py]: :+1: Thanks, @snopoke, I'll check that out.

**Comment by gcapalbo** [comment]
@kaapstorm sorry, was offline yesterday so didn't get a chance to review your updates yet, will provide feedback today

File: corehq/apps/reports/standard/sms.py (hunk 3 above)
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: cool, i didn't realize that applied to slicing too

File: corehq/apps/reports/standard/sms.py [hunk 10]
```diff
@@ -324,48 +334,141 @@ def uses_locations(self):
...
+        def get_timestamp(date_):
+            timestamp = ServerTime(date_).user_time(self.timezone).done()
+            table_cell = self._fmt_timestamp(timestamp)
+            return table_cell['html']
+
+        def get_contact_link(couch_recipient, couch_recipient_doc_type):
+            doc_info = self.get_recipient_info(couch_recipient_doc_type, couch_recipient, contact_cache)
+            table_cell = self._fmt_contact_link(couch_recipient, doc_info)
+            return table_cell['html']
+
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: I think it might be cleaner if these methods were defined on the `MessageLogReport` rather than being defined in the `rows` property

File: corehq/apps/reports/standard/sms.py (hunk 9 above)
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: I think we should try to avoid adding the values as html attributes. Also I don't think you'll have access to `cell['raw']` from `export_rows` since it just gets its data from `rows` which doesn't use table cells anymore, unless I'm missing something?

I think one way to solve this problem is just to have a method or property separate from `rows` and `export_table` that generates all of the data, including the extra columns needed for export.  Then have `rows` exclude the extra columns.

How come we removed the use of table cells, was that to reduce on memory usage?

File: corehq/apps/reports/standard/sms.py [hunk 11]
```diff
@@ -324,48 +334,141 @@ def uses_locations(self):
...
+            date__range=(startdate, enddate),
+        ).exclude(
+            # Exclude outgoing messages that have not yet been processed
+            direction=OUTGOING,
+            processed=False
+        )
+        data = filter_by_types(data)
+        data = order_by_col(data)
+        if self.pagination:
+            data = data[self.pagination.start:self.pagination.start + self.pagination.count]
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: I tested this locally and it looks like when exporting the data, it's still only exporting a single page of data rather than all of the data. I think the right way to solve this would be similar to what I mentioned in https://github.com/dimagi/commcare-hq/pull/8041/files#r38780841 - basically just have a separate method that gets all of the data which takes options pagination parameters and is called from `export_table` and `rows`, and pass it pagination parameters only when its invoked from `rows`.

File: corehq/apps/reports/standard/sms.py [hunk 12]
```diff
@@ -1,4 +1,6 @@
 import cgi
+from lxml.etree import ParseError
```
Comment (dimagimon) [inline corehq/apps/reports/standard/sms.py]: F401 'ParseError' imported but unused

File: corehq/apps/reports/standard/sms.py [hunk 13]
```diff
@@ -1,4 +1,6 @@
...
+from lxml import etree
```
Comment (dimagimon) [inline corehq/apps/reports/standard/sms.py]: F401 'etree' imported but unused

**Comment by kaapstorm** [comment]
Thanks @snopoke and @gcapalbo.  I've made the changes you suggested: Standard SMS reports now use SQL throughout, and the export returns contact info in a normal way.

File: corehq/apps/reports/standard/sms.py [hunk 14]
```diff
@@ -127,14 +120,17 @@ def _sms_count(user, startdate, enddate, message_type='SMSLog'):
...
-            startkey=[user.doc_type, user._id, message_type, direction, start_timestamp],
-            endkey=[user.doc_type, user._id, message_type, direction, end_timestamp],
-            reduce=True).all()
-        ret[direction] = results[0]['value'] if results else 0
+        count = SMS.objects.filter(
+            couch_recipient_doc_type=user.doc_type,
+            couch_recipient=user._id,
+            direction=direction,
+            date__range=(start_timestamp, end_timestamp),
+        ).count()
```
Comment (snopoke) [inline corehq/apps/reports/standard/sms.py]: Can you not do an aggregation on `direction`? Would mean you only need 1 query.

File: corehq/apps/reports/standard/sms.py [hunk 15]
```diff
@@ -322,51 +321,121 @@ def uses_locations(self):
...
+                filters = workflow_filter
+            return data_.filter(filters)
+
+        def order_by_col(data_):
+            col_fields = ['date', 'couch_recipient', 'phone_number', 'direction', 'text']
+            sort_col = self.request_params.get('iSortCol_0')
+            if sort_col is not None and sort_col < len(col_fields):
+                data_ = data_.order_by(col_fields[sort_col])
+            if self.request_params.get('sSortDir_0') == 'desc':
+                data_ = data_.reverse()
```
Comment (snopoke) [inline corehq/apps/reports/standard/sms.py]: I think this will reverse the order of the list in python, is it not better to let the DB do that?

File: corehq/apps/sms/models.py [hunk 16]
```diff
@@ -169,14 +169,22 @@ def count_outgoing_by_domain(cls, domain, start_date = None, end_date = {}):
         return 0
     
     @classmethod
-    def by_domain_date(cls, domain, start_date = None, end_date = {}):
+    def by_domain_date(cls, domain, start_date=None, end_date=None, limit=None, skip=None):
```
Comment (snopoke) [inline corehq/apps/sms/models.py]: Is this change still necessary?

Comment (kaapstorm) [inline corehq/apps/sms/models.py]: Unused now. I'll revert this change.

File: corehq/apps/reports/standard/sms.py (hunk 15 above)
Comment (kaapstorm) [inline corehq/apps/reports/standard/sms.py]: It's a [QuerySet method][1], but I'm glad you pointed it out, because it needs to come with the `order_by` call. I'll check that `sSortDir_0` won't be set without `iSortCol_0`.

  [1]: https://docs.djangoproject.com/en/1.8/ref/models/querysets/#reverse

File: corehq/apps/reports/standard/sms.py (hunk 14 above)
Comment (kaapstorm) [inline corehq/apps/reports/standard/sms.py]: :+1:

File: corehq/apps/reports/standard/sms.py [hunk 17]
```diff
@@ -113,9 +105,10 @@ def _fmt(val):
             )
         ]
 
-def _sms_count(user, startdate, enddate, message_type='SMSLog'):
+
+def _sms_count(user, startdate, enddate):
```
Comment (kaapstorm) [inline corehq/apps/reports/standard/sms.py]: Might be nice to sync `VerifiedNumber` to SQL too, and then `_sms_count` and `len(get_verified_numbers)` can be replaced with a single query ... but not in this PR I don't think.

Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: Yeah, I've got a phone number migration planned too, I'll keep this in mind when the time comes

File: corehq/apps/reports/standard/sms.py [hunk 18]
```diff
@@ -322,51 +317,121 @@ def uses_locations(self):
...
+                WORKFLOW_BROADCAST,
+                WORKFLOW_CALLBACK,
+                WORKFLOW_DEFAULT,
+                WORKFLOW_PERFORMANCE,
+            )
+            incl_survey = MessageTypeFilter.OPTION_SURVEY.lower() in filtered_types
+            incl_other = MessageTypeFilter.OPTION_OTHER.lower() in filtered_types
+            is_workflow_relevant = Q(workflow__in=relevant_workflows)
+            workflow_filter = Q(is_workflow_relevant & Q(workflow__in=filtered_types))
+            survey_filter = ~Q(xforms_session_couch_id__exact=None)
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: I think it's preferable to use `xforms_session_couch_id__isnull=False`, but not a blocker

File: corehq/apps/reports/standard/sms.py [hunk 19]
```diff
@@ -322,51 +317,121 @@ def uses_locations(self):
...
+            return table_cell['raw'] if raw else table_cell['html']
 
-            phone_number = message.phone_number
-            if abbreviate_phone_number and phone_number is not None:
-                phone_number = phone_number[0:7] if phone_number[0:1] == "+" else phone_number[0:6]
+        startdate = json_format_datetime(self.datespan.startdate_utc)
+        enddate = json_format_datetime(self.datespan.enddate_utc)
+        data = SMS.objects.filter(
+            domain=self.domain,
+            date__range=(startdate, enddate),
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: It shouldn't be necessary to json format the start and end timestamps anymore

File: corehq/apps/reports/standard/sms.py [hunk 20]
```diff
@@ -322,51 +317,121 @@ def uses_locations(self):
...
+        return self._get_rows(paginate=True, contact_info=False)
+
+    @property
+    def total_records(self):
+        startdate = json_format_datetime(self.datespan.startdate_utc)
+        enddate = json_format_datetime(self.datespan.enddate_utc)
+        count = SMS.objects.filter(
+            domain=self.domain,
+            date__range=(startdate, enddate),
+        ).count()
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: Same comment as above - would be better if we didn't json format the timestamps anymore

Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: Also, I think the total records needs to have the same filters applied because right now it shows the same count with filters applied and no filters applied.

File: corehq/apps/reports/standard/sms.py [hunk 21]
```diff
@@ -240,21 +232,24 @@ def export_table(self):
...
+
+    So, to have this report abbreviate the phone number to only the first four digits for a certain domain, add
+    the domain to the list in settings.MESSAGE_LOG_OPTIONS["abbreviated_phone_number_domains"]
+    """
     name = ugettext_noop('Message Log')
     slug = 'message_log'
+    ajax_pagination = True
 
     exportable = True
+    exportable_all = True
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: When I click the export button now, it gives me a message that the export will be emailed to me. Was that intentional, and can we keep it as a regular download?

File: corehq/apps/reports/standard/sms.py [hunk 22]
```diff
@@ -322,51 +317,121 @@ def uses_locations(self):
...
         abbreviate_phone_number = (self.domain in abbreviated_phone_number_domains)
 
         contact_cache = {}
-        message_type_filter = self.get_message_type_filter()
 
-        for message in data:
-            if message.direction == OUTGOING and not message.processed:
-                continue
+        def filter_by_types(data_):
+            filtered_types = {t.lower() for t in MessageTypeFilter.get_value(self.request, self.domain)}
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: I tested out the filters and it's not working because of converting this to lowercase. It looks like we can just remove the `.lower()` call entirely here

File: corehq/apps/reports/standard/sms.py [hunk 23]
```diff
@@ -322,51 +317,121 @@ def uses_locations(self):
...
+            relevant_workflows = (
+                WORKFLOW_REMINDER,
+                WORKFLOW_KEYWORD,
+                WORKFLOW_BROADCAST,
+                WORKFLOW_CALLBACK,
+                WORKFLOW_DEFAULT,
+                WORKFLOW_PERFORMANCE,
+            )
+            incl_survey = MessageTypeFilter.OPTION_SURVEY.lower() in filtered_types
+            incl_other = MessageTypeFilter.OPTION_OTHER.lower() in filtered_types
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: can remove `.lower()` from these two lines too

File: corehq/apps/reports/standard/sms.py [hunk 24]
```diff
@@ -322,51 +317,121 @@ def uses_locations(self):
...
+            # We can chain ANDs together, but not ORs, so we have to do all the ORs at the same time.
+            if incl_survey and incl_other:
+                filters = (workflow_filter | survey_filter | other_filter)
+            elif incl_survey:
+                filters = (workflow_filter | survey_filter)
+            elif incl_other:
+                filters = (workflow_filter | other_filter)
+            else:
+                filters = workflow_filter
+            return data_.filter(filters)
```
Comment (gcapalbo) [inline corehq/apps/reports/standard/sms.py]: Also realized the location filter is missing here, see https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/standard/sms.py#L332

**Comment by kaapstorm** [comment]
Thanks for all the feedback @gcapalbo. This adds the location filter, filters the row count, drops `.lower()` from `filter_by_types()`, and overrides `export_rows` instead of setting `exportable_all = True` so that exports are downloaded as normal.

I'd like to add some tests to this. But I'd also like to be able to close [FB 180161][1]. Do you think it's best to keep [180161][1] open and add the tests in this PR, or close [180161][1], open a new case and add the tests in a second PR? (Or does it really not make any difference, and I'm just over-thinking this?)

  [1]: http://manage.dimagi.com/default.asp?180161

**Comment by gcapalbo** [comment]
@kaapstorm Reviewed this today, looks good. I started testing it and fixed one thing (see latest commit) but otherwise so far so good, and am planning to finish up testing tomorrow.

Re: tests, I think tests would be great.  I agree it would be good to close the case, so once I finish testing this locally I'll merge and then the tests can be added in a separate PR.

**Comment by kaapstorm** [comment]
:+1: Thanks @gcapalbo.

**Comment by gcapalbo** [comment]
@kaapstorm I finally got a chance to finish testing this today. Everything looks great, I just made some minor fixes (see last 3 commits).  With that, we're all set to release this.

**Comment by kaapstorm** [comment]
\o/ Thanks so much @gcapalbo
